_posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md

```diff
@@ -0,0 +1,59 @@
+---
+title: Using pluck can save a bunch of memory
+description: ''
+layout: article
+category: ruby
+image:
+  base: '2018/using-pluck-can-save-a-bunch-of-memory'
+  alt: 'Guitarist'
+  source: 'https://unsplash.com/photos/DUrdHDn81mk'
+  credit: 'Jacek Dylag'
+---
+
+Active Record models are incredible flexible and provide a very large amount of functionality. One consequence of providing this large API of methods is that each individual Active Record object takes up a lot of space when loaded into memory.
+
+If you only use a single field, for example, the `id`s, then Rails has your back.
+
+## Instead of…
+
+...iterating over objects fully loaded into memory.
+
+```ruby
+Book.paperbacks.map { |book| book.title }
+#=> ["Eloquent Ruby", "Sapiens", "Agile Web Development With Rails"]
+Book.paperbacks.map(&:title)
+#=> ["Eloquent Ruby", "Sapiens", "Agile Web Development With Rails"]
+```
+
+
+## Use…
+
+...the `ActiveRelation#pluck` method to pull the fields directly from the database.
+
+```ruby
+Book.paperbacks.pluck(:title)
+#=> ["Eloquent Ruby", "Sapiens", "Agile Web Development With Rails"]
+```
+
+
+## But why?
+
+This is about speed and efficient memory usage.
+
+Active Record models can take up a lot of memory and this becomes particularly painful if you're operating over large collections. In the first example, each row from the database is transformed into an object and we're only picking a single field. All the extra functionality that a model has—and Rails prepares for you in memory—is unneeded.
```

Comment:
You've already said they can take up a lot of memory in the beginning. Therefore, how about rewording this to something like:

"The problem of Active Record models taking up a lot of memory becomes particularly painful if..." or "The fact that Active Record models can take up a lot of memory becomes particularly painful if...".

Reply:
*In the first example, each row from the database is transformed into an object and _yet_ we're only picking _out_/_using_ a single field.

Reply:
Add spaces around the hyphens...

Convert them to em-dashes if you care about being technically correct. :P

Reply:
They are em-dashes. Unicode. 😸

Reply:
Alrighty! 😁